### ISSUE TITLE
The performance of l2_attack on Pytorch

### ISSUE THREAD

**Issue by zjysteven** [carlini/nn_robust_attacks#23]
Hi! I am trying to reproduce your l2_attack on Pytorch to test the robustness of my network (I trained the model with Pytorch). 

But now there will be around 5 samples in each 100 samples that don't have adversarial counterparts according to the results (their l2 distance are 1e10 which is the defualt initial value set in the code).

The only two differences between my code (Pytorch version) and your code (Tensorflow version) are:
(1) the initialization of modifier. Since I couldn't find which distribution the "tf.variables_initializer" uses, so I just tried several initializers with different distribution in Pytorch, but still couldn't gurantee all adversarial samples are found.

(2) the Optimizer. I think there is subtle difference between the Adam optimizer of tf and that of Pytorch, according to my previous experience.

(Btw, I am sure I am using pre-softmax outputs instead of post-softmax ones.)

I have tried to adjust the learning rate and initial constant, but that didn't make much difference. Due to the GPU limitation, it's not practical for me to set large max_iteration & binary_search_step and correspondingly small learning rate.

So I wonder if you have any experience with this issue. I appreciate any insights or suggestions. Thanks!

**Comment by carlini** [comment]
There are basically four things that the code does:

(1) Use a hinge loss as defined in the paper. Make sure you do this PRE-SOFTMAX, otherwise things will not be correct.
(2) Use a better optimizer than standard SGD. Exactly which one shouldn't be important.
(3) Choose the lagrangian multiplier through binary search.
(4) Run for many iterations.

If you're testing on MNIST, then you should be able to set the lagrangian multiplier to 10.0 or so as a baseline. If you use a learning rate of 10^-1, then only 100 iterations should give approximately reasonable numbers. Those settings should give reasonable results if everything else is correct.

**Comment by zjysteven** [comment]
Thanks! I'm now testing on CIFAR-10, and I set learning rate: 0.1, iterations: 100, binary search step: 1. I guess I will try MNIST first and see what's going on.

**Comment by carlini** [comment]
Those settings should be reasonable, too. I would recommend verifying you are doing the pre-softmax correctly. If you do things post-softmax, results are very bad.

**Comment by carlini** [comment]
Thanks for the update. That's interesting. Adam is supposed to set the learning rate adaptively, so not sure why the pytorch one isn't doing it.
